processor/src/trace.rs

```diff
@@ -159,54 +97,36 @@ impl ExecutionTrace {
 impl Trace for ExecutionTrace {
     type BaseField = Felt;
 
-    fn width(&self) -> usize {
-        TRACE_WIDTH
+    fn layout(&self) -> &TraceLayout {
+        &self.layout
     }
 
     fn length(&self) -> usize {
-        self.system[0].len()
-    }
-
-    fn get(&self, col_idx: usize, row_idx: usize) -> Felt {
-        match col_idx {
-            i if SYS_TRACE_RANGE.contains(&i) => self.system[i - SYS_TRACE_OFFSET][row_idx],
-            i if STACK_TRACE_RANGE.contains(&i) => self.stack[i - STACK_TRACE_OFFSET][row_idx],
-            i if RANGE_CHECK_TRACE_RANGE.contains(&i) => {
-                self.range[i - RANGE_CHECK_TRACE_OFFSET][row_idx]
-            }
-            i if AUX_TRACE_RANGE.contains(&i) => self.aux_table[i - AUX_TRACE_OFFSET][row_idx],
-            _ => panic!("invalid column index"),
-        }
+        self.main_trace.num_rows()
     }
 
     fn meta(&self) -> &[u8] {
         &self.meta
     }
 
-    fn read_row_into(&self, step: usize, target: &mut [Felt]) {
-        for (i, column) in self.system.iter().enumerate() {
-            target[i + SYS_TRACE_OFFSET] = column[step];
-        }
-
-        for (i, column) in self.stack.iter().enumerate() {
-            target[i + STACK_TRACE_OFFSET] = column[step];
-        }
-
-        for (i, column) in self.range.iter().enumerate() {
-            target[i + RANGE_CHECK_TRACE_OFFSET] = column[step];
-        }
+    fn main_segment(&self) -> &Matrix<Felt> {
+        &self.main_trace
+    }
 
-        for (i, column) in self.aux_table.iter().enumerate() {
-            target[i + AUX_TRACE_OFFSET] = column[step];
-        }
+    fn build_aux_segment<E: FieldElement<BaseField = Felt>>(
+        &mut self,
+        _aux_segments: &[Matrix<E>],
+        _rand_elements: &[E],
+    ) -> Option<Matrix<E>> {
+        // TODO: implement
+        unimplemented!()
     }
 
-    fn into_columns(self) -> Vec<Vec<Felt>> {
-        let mut result: Vec<Vec<Felt>> = self.system.into();
-        self.stack.into_iter().for_each(|v| result.push(v));
-        self.range.into_iter().for_each(|v| result.push(v));
-        self.aux_table.into_iter().for_each(|v| result.push(v));
-        result
+    fn read_main_frame(&self, row_idx: usize, frame: &mut EvaluationFrame<Felt>) {
+        let next_row_idx = (row_idx + 1) % self.length();
```

Comment:
Why is this  mod `self.length()`? If we get a row index where this has an effect, why isn't that a situation where we would want an error?

Reply:
When we read an evaluation frame for the last row, the `next` row is actually the first row in the trace (we wrap around). The `mod` is here to handle this situation.

Another note, transition constraints between the last and the first row of the trace would not hold, but this is OK because the domain for transition constraints is set to be all rows except for the last one.

Reply:
Ok, I see. It does make sense, but I hadn't been thinking about wrapping around, mainly because of this: "transition constraints between the last and the first row of the trace would not hold". I think this might warrant a comment, as it feels a bit counterintuitive, but that could just be me :)